pvlib/iotools/cams.py

```diff
@@ -0,0 +1,357 @@
+    DHI, dhi*                float   Diffuse horizontal radiation
+    BNI, dni*                float   Beam (direct) radiation normal to the sun
+    Reliability*             float   Fraction of reliable data in summarization
+    =======================  ======  ==========================================
+
+    *Parameters only returned if service='cams_radiation'. For description of
+    additional output parameters in verbose mode, see [1]_ and [2]_.
+
+    Note that it is recommended to specify the latitude and longitude to at
+    least the fourth decimal place.
+
+    Variables corresponding to standard pvlib variables are renamed,
+    e.g. `sza` becomes `solar_zenith`. See the
+    `pvlib.iotools.cams.CAMS_RADIATION_VARIABLE_MAP` dict for the complete
+    mapping.
+
+    See Also
+    --------
+    pvlib.iotools.read_cams_radiation, pvlib.iotools.parse_cams_radiation
+
+    Raises
+    ------
+    requests.HTTPError
+        If the request is invalid, then an XML file is returned by the CAMS
+        service and the error message will be raised as an expcetion.
+
+    References
+    ----------
+    .. [1] `CAMS Radiation Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-radiation-service/info>`_
+    .. [2] `CAMS McClear Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-mcclear/info>`_
+    .. [3] `CAMS McClear Automatic Access
+       <http://www.soda-pro.com/help/cams-services/cams-mcclear-service/automatic-access>`_
+    """
+    if time_step in TIME_STEPS_MAP.keys():
+        time_step_str = TIME_STEPS_MAP[time_step]
+    else:
+        warnings.warn('Time step not recognized, 1 hour time step used!')
+        time_step, time_step_str = '1h', 'PT01H'
+
+    if (verbose is True) & ((time_step != '1min') | (time_ref != 'UT')):
+        verbose = False
+        warnings.warn("Verbose mode only supports 1 min. UT time series!")
+
+    # Format verbose variable to the required format: {'true', 'false'}
+    verbose = str(verbose).lower()
+
+    if altitude is None:  # Let SoDa get elevation from the NASA SRTM database
+        altitude = -999
+
+    # Start and end date should be in the format: yyyy-mm-dd
+    start_date = start_date.strftime('%Y-%m-%d')
+    end_date = end_date.strftime('%Y-%m-%d')
+
+    email = email.replace('@', '%2540')  # Format email address
+    service = 'get_{}'.format(service.lower())  # Format CAMS service string
+
+    # Manual format the request url, due to uncommon usage of & and ; in url
+    url = ("http://{}/service/wps?Service=WPS&Request=Execute&"
+           "Identifier={}&version=1.0.0&RawDataOutput=irradiation&"
+           "DataInputs=latitude={};longitude={};altitude={};"
+           "date_begin={};date_end={};time_ref={};summarization={};"
+           "username={};verbose={}"
+           ).format(server, service, latitude, longitude, altitude, start_date,
+                    end_date, time_ref, time_step_str, email, verbose)
```

Comment:
can you reformat into one or more f-strings? It's hard to keep track of this many parameters passed to format.

That format is a bit of a pain. What about something like

```python
base_url = f"http://{server}/service/wps"
data_inputs = (
    f"latitude={latitude};longitude={longitude};altitude={altitude};"
    f"date_begin={start_date};date_end={end_date};time_ref={time_ref};"
    f"summarization={time_step_str};username={email};verbose={verbose}"
params = {
    'Service': WPS,
    'Request': 'Execute',
    'Identifier': service,
    'version': 1.0.0,
    'RawDataOutput': 'irradiation',
    'DataInputs': data_inputs
}

res = requests.get(base_url, params=params)
```

Reply:
@wholmgren Unfortunately, this does not work, even though it is indeed much cleaner and understandable. When the request.get function formats the URL, it converts any equal signs in the parameter values to %3D and semi-colons to %3B. I found this by looking at the generated URL (res.url).

E.g. the first part of the data input section of the URL needs to be "&DataInputs=latitude=2.9;longitude=-54.0;altitude=-999;", but it is instead converted by the request.get function into "&DataInputs=latitude%3D2.9%3Blongitude%3D-54.0%3Baltitude%3D-999%3B" which is not recognized by the API.

However, adding the datainputs manually to the base URL and letting request.get format part of it works:

    base_url = f"http://{server}/service/wps"

    data_inputs = (
        f"latitude={latitude};longitude={longitude};altitude={altitude};"
        f"date_begin={start_date};date_end={end_date};time_ref={time_ref};"
        f"summarization={time_step_str};username={email};verbose={verbose}")

    params = {
        'Service': 'WPS',
        'Request': 'Execute',
        'Identifier': service,
        'version': '1.0.0',
        'RawDataOutput': 'irradiation',
        ####'DataInputs': data_inputs ####  this line will be deleted
        }

    res = requests.get(base_url + '?DataInputs=' + data_inputs, params=params)

What do you think of that?

Reply:
looks like a good solution

Reply:
This has been done in the new commit "Reworked URL formatting & comments on HTTP status"